New post: The evolution of a Dask Distributed user (#99)
dask/dask-blog: _posts/2021-05-21-dask-distributed-user-journey.md

---
layout: post
title: The evolution of a Dask Distributed user
tagline: The critical path to scaling from one machine to thousands
author: Jacob Tomlinson (NVIDIA)
tags: [Distributed, Tools, Organisations]
theme: twitter
---

{% include JB/setup %}

> This week was the 2021 Dask Summit and [one of the workshops](https://summit.dask.org/schedule/presentation/20/deploying-dask/) that we ran covered many deployment options for Dask Distributed.

Reviewer: It'd be great to also add a link to the workshop video recording (I know it's not up on youtube yet, but will be soon).

We covered local deployments, SSH, Hadoop, Kubernetes, the Cloud and managed services, but one question that came up a few times was "where do I start?".

I wanted to share the journey that I've seen many Dask users take in the hopes that you may recognize yourself as being somewhere along this path and it may inform you where to look next.

## In the begnning

As a user who is new to Dask you're likely working your way through [the documentation](https://docs.dask.org/en/latest/index.html) or perhaps [a tutorial](https://github.com/dask/dask-tutorial).

We often introduce the concept of the distributed scheduler early on, but you don't need it to get initial benefits from Dask. Switching from Pandas to Dask for larger than memory datasets is a common entry point and performs perfectly well using the default threaded scheduler.

```python
# Switching from this
import pandas as pd
df = pd.read_csv('/data/.../2018-*-*.csv')

> df.groupby(df.account_id).balance.sum().compute()

Reviewer 2: Spurious `.compute()` call for the pandas dataframe, surely?

# To this
import dask.dataframe as dd
df = dd.read_csv('/data/.../2018-*-*.csv')
df.groupby(df.account_id).balance.sum().compute()
```

But by the time you're a few pages into the documentation you're already being encouraged to create `Client()` and `LocalCluster()` objects.

*__Note__: When you create a `Client()` with no arguments/config set Dask will launch a `LocalCluster()` object for you under the hood. So often `Client()` is equivalent to `Client(LocalCluster())`.*

This is a common area for users to stick around in, launch a local distributed scheduler and do your work maximising the resources on your local machine.

```python
from dask.distributed import Client
import dask.dataframe as dd

client = Client()

df = dd.read_csv('/data/.../2018-*-*.csv')
df.groupby(df.account_id).balance.sum().compute()
```

## Breaking free from your machine

> Once you get used to task graphs and work scheduling you may begin thinking about how you can expand your computation beyond your computer.

Reviewer 2: There's a bit of repetition with 'computation' and the 'computer'. Given you used 'local machine' above, you could use the same terminology here to avoid that repetition.

With Dask Distributed you can start to expand your workload onto these machines using `SSHCluster`. All you need is your SSH keys set up so you can log into those machines without a password.

```python
from dask.distributed import Client, SSHCluster
import dask.dataframe as dd

cluster = SSHCluster(
    [
        "localhost",
        "alices-desktop.lan",
        "bobs-desktop.lan",
        "team-server.lan",
    ]
)
client = Client(cluster)

df = dd.read_csv("/data/.../2018-*-*.csv")
df.groupby(df.account_id).balance.sum().compute()
```

Now the same workload can run on all of the CPUs in our little ad-hoc cluster, using all the memory and pulling data from the same shared storage.

## Moving to a compute platform

Using (and abusing) hardware like desktops and shared servers will get you reasonably far, but probably to the dismay of your IT team.

Organisations who have many users trying to perform large compute workloads will probably be thinking about or already have some kind of platform that is designated for running this work.

> The platforms your organisation has will be the result of many arbitrary technology choices. What programming languages does your company use? What deals did vendors offer at the time of procurement? What skills do the current IT staff have? What did your CTO have for breakfast the day they chose a vendor?

Reviewer: Maybe "sometimes arbitrary" would be better here? Still cheeky, but leaves room to be less potentially insulting. 

Often a lot of the constraints that lead to particular decisions aren't known to users, so things can feel more arbitrary than they actually are.

Author: Yeah I know what you mean. I'm not sure "arbitrary" is necessarily the right word. But I couldn't think of a better one.

Folks make different decisions for various reasons. All reasons are generally valid. All choices are usually similar.

Decisions like Dell or HP for servers, AWS or GCP for cloud and Cray or IBM for HPC are usually not made because one technology is vastly superior to the other. They are made because there may be some small detail about your org which makes one choice slightly more compelling. These choices have a big impact on the details that engineers/scientists/etc have to care about, but very small impact in the grand scheme of things.

With Dask tooling we are attempting to homogenise those details that folks have to care about, so that the decision becomes "to Dask or not to Dask?" rather than "can I even Dask given our past technology decisions?".

Author: Even after writing all that I'm still not sure if I can think of a better word than "arbitrary".

> Within Dask we don't care which platform decisions your organisations make and try to build deployment tools for as many popular platforms as we can including:

Reviewer: ```suggestion
At Dask we don't care which platform decisions your organisations make. We try to build deployment tools for as many popular platforms as we can including:
```

Reviewer 2: I must say I prefer 'within' over 'at' here, because:
* 'within' is often used for contents of codebases / libraries: "within this library you'll find functions to do x and y..."
* 'at' sounds more like a company: "at Dask Inc we enable you to do x and y..."

With this approach your IT team has full control over the compute resource. They can ensure folks get their fair share with quotas and queues. But you as a user gets the same Dask experience you are used to on your local machine.

Your data may be in a slightly different place on these platforms though. Perhaps you are on the cloud and your data is in object storage for example. Thanks to tools built on [fsspec](https://filesystem-spec.readthedocs.io/en/latest/) like [s3fs](https://github.com/dask/s3fs) or [adlfs](https://pypi.org/project/adlfs/) we can read this data in pretty much the same way. So still not much change to your workflow.

```python
from dask.distributed import Client
from dask_cloudprovider.azure import AzureVMCluster
import dask.dataframe as dd

cluster = AzureVMCluster(resource_group="<resource group>",
                         vnet="<vnet>",
                         security_group="<security group>",
                         n_workers=1)
client = Client(cluster)

df = dd.read_csv("adl://.../2018-*-*.csv")
df.groupby(df.account_id).balance.sum().compute()
```

## Centralizing your Dask resources

> When your organisation gets enough folks adopting and using Dask it may be time for your IT team to step in and provide you with a managed service. Having many users submitting many ad-hoc clusters in a myriad of ways is likely to be less efficient than a centralised managed and more importantly ordained service from IT.

Reviewer: I don't think it's clear here about why the alternative approach is less efficient in several ways, or why it's important to have it as an official IT service.

I'd like to see a suggestion on where the threshold is for moving from `dask-jobqueue`/`dask-foo`/etc. to Dask Gateway. That way users can be confident about which "tier" of dask deployment options suits them at the moment.

Reviewer 2: And maybe change 'centralised managed' to one of:
* 'centralised, managed' or
* 'centrally managed'

> One approach they can take is to deploy [Dask Gatweay](https://gateway.dask.org/). This can be deployed by an administrator and provides a central hub which launches Dask clusters on behalf of users. It supports many types of authentication so it can hook into whatever your organisation uses and supports many of the same backend compute platforms that the standalone tools do, including Kubernetes, Hadoop and HPC.

Reviewer 2: Typo: 'gatweay' --> 'gateway'

Users will need to authenticate with the gateway, but then can launch Dask clusters in a platform agnostic way.

```python
from dask.distributed import Client
from dask_gateway import Gateway
import dask.dataframe as dd

gateway = Gateway(
    address="http://daskgateway.myorg.com",
    auth="kerberos"
)
cluster = gateway.new_cluster()
client = Client(cluster)

df = dd.read_csv("/data/.../2018-*-*.csv")
df.groupby(df.account_id).balance.sum().compute()
```

Again reading your data depends on how it is stored on the compute platform you are using, but the changes required are minimal.

## Managed services

If your organisation is too small to have an IT team to manage this for you, or you just have a preference for managed services, there are startups popping up to provide this to you as a service including [Coiled](https://coiled.io/) and [Saturn Cloud](https://www.saturncloud.io/s/home/).

> The large cloud vendors also have managed data science platforms including [AWS Sagemaker](https://aws.amazon.com/sagemaker/), [Azure Machine Learning](https://azure.microsoft.com/en-gb/services/machine-learning/) and [Google Cloud AI Platform](https://cloud.google.com/vertex-ai).

Reviewer 2: Taken on its own, this sentence could imply that these managed data science platforms will give you dask straight out the box. Taken with the following sentence it becomes more clear that you don't, but it might be worth making it clear that you as the user will still need to set up dask resources when on the cloud, just as you will locally / on a cluster.

> While the cloud services are focussed on batch processing and machine learning today these companies also have managed services for Spark and with Dask's increasing popularity it wouldn't surprise me if more managed Dask services appear in the years to follow.

Reviewer 2: ```suggestion
While the cloud services are focussed on batch processing and machine learning today, these companies also have managed services for Spark. With Dask's increasing popularity it wouldn't surprise me if more managed Dask services are released by these cloud vendors in the years to follow.
```

☝️ I think these changes might make your point here a bit more clear. Do you think it might be worth expanding on the concept of a 'managed service' here? Or are you happy that your readers are likely to be confident with the concept of managed services? I've also tried to clarify that you reckon the cloud vendors will release managed dask services in the near future.

Reviewer 2: Given my previous comment as well, you may also want to make it clear that this is aspirational, rather than the current state of affairs..?

## Summary

One of the most powerful features of Dask is that your code can stay pretty much the same regardless of how big or complex the distributed compute cluster is. It scales from a single machine to thousands of servers with ease.

But scaling up requires both user and organisational growth and folks already seem to be treading a common path on that journey.

> Hopefully this post will give you an idea of where you are on that path and where to jump to next. Whether you're new to the community and discovering the power of multi-core computing or an hold hand who is trying to wrangle hundreds of users who all love Dask, good luck!

Reviewer 2: 'an hold hand' --> 'an old hand' (😁)
